### DISCUSSION TITLE
How do I make completions on <Tab>?

### DISCUSSION THREAD

**Discussion by justjew** [LazyVim/LazyVim#39] [Q&A] [answered]
I tried to override `nvim-cmp` under `lua/plugins` like this:
```lua
  {
    "hrsh7th/nvim-cmp",
    opts = function()
      local cmp = require("cmp")
      return {
        mapping = cmp.mapping.preset.insert({
          ["<tab>"] = cmp.mapping.confirm({ select = true }),
        }),
      }
    end
  },
```
Not working.

Then I tried to set a keymap like this:
```lua
local cmp = require("cmp")
vim.keymap.set("s", "<tab>", function()
  cmp.mapping.confirm({ select = true })
end)
```
Also not working.

Pressing Enter to autocomplete makes me cry. I don't want to cry.

**Comment by appelgriebsch** [comment]
Have a look here: https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/coding.lua#L24

```lua
   opts = function(_, opts)
      local cmp = require("cmp")
      opts.mapping = cmp.mapping.preset.insert({
        ["<Down>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Select }),
        ["<Up>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Select }),
        ["<C-j>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Select }),
        ["<C-k>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Select }),
        ["<C-b>"] = cmp.mapping(cmp.mapping.scroll_docs(-1), { "i", "c" }),
        ["<C-f>"] = cmp.mapping(cmp.mapping.scroll_docs(1), { "i", "c" }),
        ["<C-Space>"] = cmp.mapping(cmp.mapping.complete(), { "i", "c" }),
        ["<C-e>"] = cmp.mapping {
          i = cmp.mapping.abort(),
          c = cmp.mapping.close(),
        },
        -- Accept currently selected item. If none selected, `select` first item.
        -- Set `select` to `false` to only confirm explicitly selected items.
        ["<CR>"] = cmp.mapping.confirm { select = true },
      })
```

works for me...

**Reply by danielo515** [reply]
> ```lua
>    opts = function(_, opts)
>       local cmp = require("cmp")
> ```

Whyou require cmp? Isn't the first argument of the function supposed to be the plugin itself?

**Reply by rockfruit** [reply]
I can't believe how dumb I had to feel to finally get this working.  Thank you @appelgriebsch

**Reply by Nikola-Milovic** [reply]
Is it possible to override just a portion of the keybindings, not all of them?

**Comment by al-ce** [comment]
Unsure if this is related to #44 but setting `<tab>` for cmp completion (whether with the `opts.mapping` method suggested above or the method used in the linked issue), I get the following behavior:

1) start nvim, check lazy, all cmp plugins are loaded
2) enter insert mode, tab works as intended with indentation, but cmp competion doesn't show results. Snippets do, but tab no <CR> complete them
3) Go to normal mode, reenter insert. Now cmp popups show, <tab> completes them, but
4) Pressing `<tab>` when no completion pop up shows results in `<tab>` as text being inserted into the buffer, rather than indentation.

So, `i<tab><esc>i<tab>` would produce this line:
`    <tab>`

`:map <tab>` before entering insert for the firt time
`map <tab> s  <Tab>       * <Lua 105: ~/.local/share/nvim/lazy/lazy.nvim/lua/lazy/core/handler/keys.lua:92>`

after:
`map <tab> s  <Tab>       * <Lua 123: ~/.local/share/nvim/lazy/LazyVim/lua/lazyvim/plugins/coding.lua:25>`

I can't find what's going on with my config that might be doing it, but I'm assuming it's a me thing. Just posting this here in case anyone has any suggestions for where to look!

**Reply by MarWeUMR** [reply]
Which `nvim-cmp` version / commit are you using? The literal `<tab>` insertion happend for me with `nvim-cmp v0.0.1` which was loaded by default from `lazy.nvim`. I had to specify `version=false` to resolve that, but this only happens with my actual config, not when I use the minimal repro from #44, so I figured that was some issue on my side.

**Comment by folke** [comment, marked as answer]
On popular request, I've added an [example](https://github.com/LazyVim/starter/blob/888600e7ff4b0d8be0db18db932632830bfb7804/lua/plugins/example.lua#L218) on how to setup **supertab** behavior with `cmp` and `luasnip`.

Make sure you updated `lazy.nvim` to the latest version to make this work! (specifically `keys=` as function is new)

**Reply by folke** [reply]
Just noticed this is not actually what op wants, but the spec is similar, just change tab to confirm and make sure you disable the luasnip keys handler

**Reply by al-ce** [reply]
Thank you so much!

**Reply by jcontonio** [reply]
@folke I've adding the following to plugins/coding.lua:

```
return {
  -- Use <tab> for completion and snippets (supertab)
  -- first: disable default <tab> and <s-tab> behavior in LuaSnip
  {
    "L3MON4D3/LuaSnip",
    keys = function()
      return {}
    end,
  },
  -- then: setup supertab in cmp
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      "hrsh7th/cmp-emoji",
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      local has_words_before = function()
        unpack = unpack or table.unpack
        local line, col = unpack(vim.api.nvim_win_get_cursor(0))
        return col ~= 0 and vim.api.nvim_buf_get_lines(0, line - 1, line, true)[1]:sub(col, col):match("%s") == nil
      end

      local luasnip = require("luasnip")
      local cmp = require("cmp")

      opts.mapping = vim.tbl_extend("force", opts.mapping, {
        ["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_next_item()
            -- You could replace the expand_or_jumpable() calls with expand_or_locally_jumpable()
            -- they way you will only jump inside the snippet region
          elseif luasnip.expand_or_jumpable() then
            luasnip.expand_or_jump()
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
        ["<S-Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif luasnip.jumpable(-1) then
            luasnip.jump(-1)
          else
            fallback()
          end
        end, { "i", "s" }),
      })
    end,
  },
}
```
After reloading, hitting <Tab> in an autocomplete selects a random entry in the list. Hitting <Cr> still selects the first highlighted item. Am I doing this wrong or is there a way to disable <Cr> in autocomplete? It's rather hectic having autocomplete popup all the time when typing regular text.

**Reply by OracleMachine** [reply]
This solution works *almost* perfect but with only one catch. The first time you tap <TAB>, it didnt insert with the first item in completion list. It skip and insert the second item.

**Reply by dpetka2001** [reply]
I'm guessing it's selecting the second item because of `cmp.select_next_item()` when pressing `TAB`, but the default behavior is that when you start autocompleting it already selects the first item by default, so hitting `TAB` goes to the next item which is the second. If you want the first item you should just press `Enter` and press `TAB` only if you want to go to further items down the completion list.
Or you can you try the following in your cmp spec:
```lua
      opts.preselect = cmp.PreselectMode.None
      opts.completion = {
        completeopt = "noselect",
      }
```
Just put it before `opts.mapping`.  Of course, that means, that with the above code nothing would be selected by default and you would have to press `TAB` to select something.

**Reply by marcocamilo** [reply]
Since with `unpack(vim.api.nvim_win_get_cursor(0))` I got the diagnostic `Deprecated. (Defined in Lua 5.1/LuaJIT, current is Lua 5.4 .)`, I found that replacing the unpack function with `table.unpack` removed the diagnostic. 

But then I get `E5108: Error executing lua: /Users/marco-andres/.config/nvim/lua/marco/plugins/cmp.lua:20: attempt to call field 'unpack' (a nil value) stack traceback:`. Any idea how to remove the diagnostic and keep the code snippet working?

**Reply by carlinhoshk** [reply]
> On popular request, I've added an [example](https://github.com/LazyVim/starter/blob/888600e7ff4b0d8be0db18db932632830bfb7804/lua/plugins/example.lua#L218) on how to setup **supertab** behavior with `cmp` and `luasnip`.
> 
> Make sure you updated `lazy.nvim` to the latest version to make this work! (specifically `keys=` as function is new)

I copy and paste the entire example in my example.lua but not working
I removed some options from lazy extra, could this be affecting it? I have already disabled everything I installed in extra, but when I change it, only enter works and never tab.

**Reply by brettc** [reply]
lazyvim uses blink now. So look here: https://cmp.saghen.dev/installation

**Reply by allenh9999** [reply]
Thanks @brettc, your comment helped me a lot!
For anyone wondering how to do it with Blink, there's a [preset](https://cmp.saghen.dev/configuration/keymap.html#super-tab) called `super-tab`, or alternatively you can do what I did to make tab autocomplete work:
```lua
"saghen/blink.cmp",
  opts = {
    keymap = {
      ["<Tab>"] = {
        function(cmp)
          if cmp.snippet_active() then
            return cmp.accept()
          else
            return cmp.select_and_accept()
          end
        end,
        "snippet_forward",
        "fallback",
      },
    },
  },
```

**Comment by neuromaancer** [comment]
There is another problem I found. I am a python user, but terribly, I use the tab for the indent. Every time if I need to indent in the new line, the completion comes out, and if I tap second, it will choose the suggestion for me. I need to press the to abort the completion first, then press for the indent. Is there any way to avoid this?

**Reply by dpetka2001** [reply]
No, because of this
```lua
["<Tab>"] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_next_item()
            -- You could replace the expand_or_jumpable() calls with expand_or_locally_jumpable()
            -- they way you will only jump inside the snippet region
          elseif luasnip.expand_or_jumpable() then
            luasnip.expand_or_jump()
          elseif has_words_before() then
            cmp.complete()
          else
            fallback()
          end
        end, { "i", "s" }),
```
It checks to see if `cmp_visible` to select the next item and it only falls back to the default as the last option. That means `cmp` completion menu must be closed and there shouldn't be any Luasnip snippet that can be expanded or jump to the next node. You can't have both (at least to my knowledge, if someone knows a better solution plz correct me if I'm wrong). You can also disable the automatic completion and just press `<c-space>` to manually invoke it when you want with the following
```lua
  {
    "nvim-cmp",
    opts = {
      completion = {
        autocomplete = false,
      },
    },
  },
```
Take your pick.

**Reply by neuromaancer** [reply]
Yeah, I saw the same. Because of this, I might go back to the old setting without using super-Tab

**Reply by BVegNow** [reply]
does this affect things like accepting autocomplete from AI suggestions? As I cannot get tab to work for those, the suggestion just appears and moves along as I tab, wonder if there are conflicts to look out for

**Comment by zihao-liu-qs** [comment]
I finally find the ultimate solution:
create file ~/.config/nvim/lua/plugins/tab-autocomplete.lua
the file content is 
```
-- lua/plugins/tab-autocomplete.lua
return {
  {
    "saghen/blink.cmp",
    opts = function(_, opts)
      opts.keymap = {
        preset = "super-tab",
        ["<Tab>"] = { "select_and_accept" },
        ["<S-Tab>"] = { "select_prev" },
      }
    end,
  },
}

```
